Use new settings system for plugin

mizaki opened the pull request:
Currently this doesn't update the current talker object but otherwise how is this?

mizaki committed:
Use new settings system for plugin

mizaki committed:
Use setting option for talker selection

lordwelch commented:
I'm not going to review this unless the tests affected by this PR are passing. If there is a particular test that you can't get to pass give me a mention and I will look at it.

mizaki committed:
Fix tests. Probably not the correct way to do this?

mizaki commented:
I've added a fix for the tests but it may be in the wrong place and the wrong way to do it too? 

It brings back the question of a new round of tests for the `talkerapi` but it is still probably better to wait until it is more finalised.

mizaki committed:
Load plugins in init.

lordwelch commented on comictaggerlib/ctoptions/file.py:
```diff
@@ -257,13 +239,17 @@ def validate_settings(options: settngs.Config[settngs.Values], parser: settngs.M
     parser.add_group("dialog", dialog, False)
     parser.add_group("filename", filename, False)
-    parser.add_group("comicvine", comicvine, False)
+    parser.add_group("talkers_general", talkers_general, False)
```
This can just be changed to "talkers" we don't need the extra long name

lordwelch commented on comictaggerlib/ctoptions/file.py:
```diff
@@ -257,13 +239,17 @@ def validate_settings(options: settngs.Config[settngs.Values], parser: settngs.M
+
+    # Register talker plugin settings
+    for talker, cls in talkers.items():
+        parser.add_group(talker, cls.comic_settings, False)
```
This needs to happen elsewhere.
This file originally only contained settings that were saved in the config file, now that it's a simple boolean that changes that it no longer strictly adheres to that but there's no reason for plugins to be passed in here.

lordwelch commented on comictaggerlib/main.py:
```diff
@@ -87,7 +88,7 @@ def register_options(self) -> None:
         )
         ctoptions.register_commandline(self.manager)
-        ctoptions.register_settings(self.manager)
+        ctoptions.register_settings(self.manager, self.talker_plugins)
```
To follow the comment in file.py don't shoehorn the talker plugins into `register_settings` create a new function `register_talker_settings`

lordwelch commented on comictaggerlib/main.py:
```diff
@@ -99,6 +100,18 @@ def parse_options(self, config_paths: ctoptions.ComicTaggerPaths) -> None:
 
+        # parse talker settings
+        for loaded in self.talker_plugins.values():
+            parse_options = getattr(loaded, "parse_settings", None)
```
what's with the naming convention switch?

lordwelch reviewed (changes requested):
If we reduce the `__init__` to:
```python
def __init__(self, version: str, cache_folder: pathlib.Path) -> None:
```

That makes it so that we can always create the object, and it should always succeed in a proper environment, so we can immediately create the object eg `talker(version, cache_folder)`.

We can even pass the `version` and `cache_folder` into `get_talkers` and then loop over all of them:
```python
def get_talkers(version, cache_folder):
  talkers = {}
  for talker in [comictalker.talkers.comicvine.ComicVineTalker]:
    try:
      object = talker(version, cache_folder)
      talkers[object.name] = object
    except Exception:
      logger.exception("Failed to load talker")
  return talkers
```
I would probably create a separate function so that people can easily get the class directly.
(though they can always use `object.__class__` for that)

Then we can make a configure (or parse_settings, etc...) method:
```python
def configure(self, config: dict[str, Any]) -> None:
```

I know flake8 used class methods but those plugins are also for the most part single function, they are given file content (as an ast or something else) and process it and spit out the answer. With the Talkers we are wrapping an entire api that we have to call multiple times to get anything useful and we may change it's state in between function calls. So I would rather keep all of the state on the instance (in `__init__`) than using class variables.

mizaki committed:
Rename talkers_general to talkers. Moved plugin option register to own file. Due to chicken and egg, first get talker classes then create objects.

mizaki commented:
With https://github.com/comictagger/comictagger/pull/419/commits/0d69ba3c494a520304632f8b404357867fe4ee9d I should have addressed all the issues. I found a bit of a chicken and egg situation and decided on the approach here given your previous comments. I'll give a brief explanation:

Because of the `init` requirement of a cache path to create a talker instance the use of a class method was required to register the talker(s) options. So, to begin with all talker classes are gathered to be passed to `ctoptions.register_talker_settings` to generate the options. After the options have been parsed, `initialize_talkers` is called and `self.plugins` (in main) is propagated with the obj in addition to the class. (It may make more sense to just overwrite the class with the object and get back to a simple dict of talkers again?)

When the object is created `set_settings` (needs a name change?) is run on the talker setting the settings. If that fails, the defaults will remain. It's possible it should hard fail (and if the GUI will be used enable the users to go to settings to pick a different talker)?

The other option as I see it is the remove the cache requirement from `init` and then the objects can be created before `register_options` is called and then `initialize_talkers` can be used to set the cache along with all the other settings.

Hope all that makes sense!

mizaki commented:
Another thing to decide is which options should be talker settings and which should be global.

`series-match-search-thresh` - I'm leaning more global. I can see how the user may want one talker to be more forgiving than another though.

`clear-metadata`, `sort-series-by-year`, `exact-series-matches-first`, `clear-form-before-populating`, 
`use-series-start-as-volume` - All same as above.

lordwelch commented:
> Because of the `init` requirement of a cache path to create a talker instance the use of a class method was required to register the talker(s) options.

opts.config.user_cache_dir
here's a patch to move initialize_dirs up a 2 lines [Patch.patch](https://github.com/comictagger/comictagger/files/10502950/Patch.patch)

> When the object is created `set_settings` (needs a name change?) is run on the talker setting the settings. If that fails, the defaults will remain. It's possible it should hard fail (and if the GUI will be used enable the users to go to settings to pick a different talker)?

Hard fail, put it in the log, the developer of the talker can fix it.

> Another thing to decide is which options should be talker settings and which should be global.
> 
> `series-match-search-thresh` - I'm leaning more global. I can see how the user may want one talker to be more forgiving than another though.
> 
> `clear-metadata`, `sort-series-by-year`, `exact-series-matches-first`, `clear-form-before-populating`, `use-series-start-as-volume` - All same as above.

Currently `series-match-search-thresh` and `use-series-start-as-volume` must be implemented by the talker. The others the Talker can't control so I don't see why they would be a per-talker option. If it proves to be useful we can change it in a separate PR

mizaki committed:
Create simple dict of talkers with objects. Moved thresh setting back to talkers (general) as it is called outside of talker.

mizaki commented:
Turns out `series-match-search-thresh` is used outside of the talker so I had to put it back in the `talkers` options. Similar to the others, as you said, if people want it, it can be done later.

Now talker instance(s) are created and `initialize_talkers` only applies the settings and will cause an exit if it fails.

lordwelch commented on comictaggerlib/main.py:
```diff
@@ -66,12 +66,15 @@ def __init__(self) -> None:
         self.register_options()
         self.parse_options(opts.config)
-        self.initialize_dirs()
+        self.initialize_talkers()
```
Move this to `self.parse_options` it should be called directly before `self.manager.get_namespace` and each talker should get the dictionary referring to just the settings they defined

lordwelch commented on comictalker/comictalkerapi.py:
```diff
@@ -15,24 +15,36 @@
+        raise TalkerError(source=talker.source_details.name, code=4, desc="Could not apply talker settings")
+
+
+def get_talkers(version: str, cache: pathlib.Path) -> dict[str, Any]:
```
`Any` should be `ComicTalker`

mizaki replied:
I had to change this to Mapping because mypy and `"Dict" is invariant`. Is there a better way?

lordwelch replied:
The `talkers` variable on L41 needed to be typed as well
`talkers: dict[str, ComicTalker] = {}`

lordwelch commented on comictalker/talkerbase.py:
```diff
@@ -137,27 +139,20 @@ def __init__(self, source: str = "", sub_code: int = 0, desc: str = "") -> None:
-        if tmp_url.path and tmp_url.path[-1] != "/":
-            tmp_url = tmp_url._replace(path=tmp_url.path + "/")
+    def comic_settings(self, parser: settngs.Manager) -> None:
+        """Talker settings."""
```
This should be `register_settings`.
docstring should be along the lines of

> Allows registering settings using the settngs package using a argparse like interface

lordwelch commented on comictalker/talkerbase.py:
```diff
@@ -137,27 +139,20 @@ def __init__(self, source: str = "", sub_code: int = 0, desc: str = "") -> None:
+        """Talker settings."""
 
-        self.api_url = tmp_url.geturl()
+    def set_settings(self, settings: argparse.Namespace) -> None:
+        """Apply talker settings from config to object."""
```
This should be `def parse_settings(self, settings: dict[str, Any])`

>  `settings` is a dictionary of options defined in `register_settings`.
> It is only guaranteed that all settings defined in `register_settings` will be present, there may be other options present in the dictionary.

lordwelch commented on comictalker/talkers/comicvine.py:
```diff
@@ -176,19 +169,66 @@ def __init__(
-        self.series_match_thresh: int = series_match_thresh
+    def set_settings(self, settings: argparse.Namespace) -> None:
+        """Set settings."""
+        if settings.comicvine_cv_remove_html_tables:
+            self.remove_html_tables = bool(settings.comicvine_cv_remove_html_tables)
```
https://www.youtube.com/watch?v=TJ_aGY6iyZY

mizaki replied:
As any bool setting should have already been checked, I changed it to just equalling the value.

lordwelch commented on comictalker/talkers/comicvine.py:
```diff
@@ -176,19 +169,66 @@ def __init__(
+                    tmp_url = tmp_url._replace(path=tmp_url.path + "/")
+
+                self.api_url = tmp_url.geturl()
+            except Exception:
+                logger.exception("Failed to parse new talker URL for %s, will use default", self.source_name)
```
This should not be caught. If there is an error it should be handled in ComicTagger.

mizaki committed:
Rename new settings talker methods. Move parse_settings for talkers to earlier and only pass talkers own settings.

mizaki committed:
typed talkers var

lordwelch commented on comictalker/comictalkerapi.py:
```diff
@@ -16,23 +16,36 @@
-def get_talkers() -> dict[str, type[ComicTalker]]:
-    """Returns all comic talker modules NOT objects"""
-    return {"comicvine": comictalker.talkers.comicvine.ComicVineTalker}
+def set_talker_settings(talker: ComicTalker, settings: dict[str, Any]) -> None:
+    try:
+        talker.parse_settings(settings)
+    except Exception as e:
+        logger.exception(
+            f"Failed to set talker settings for {talker.source_details.name}, will use defaults. Error: {str(e)}",
+        )
+        raise TalkerError(source=talker.source_details.name, code=4, desc="Could not apply talker settings")
```
What's the point of this function?
Currently the only place this is used does this
```python
try:
    for talker_name, talker in self.talker_plugins.items():
        ct_api.set_talker_settings(talker, self.options[0][talker_name])
except Exception as e:
    # Remove talker as we failed to apply the settings
    del self.talker_plugins[e.source]  # type: ignore[attr-defined]
    logger.exception("Failed to initialize talker settings. Error %s", str(e))
```
Which does not benefit from anything that this function does and catches the wrong exceptions and only removes the first talker that fails.

It should be:
```python
for talker_name, talker in list(self.talker_plugins.items()):
    try:
        ct_api.set_talker_settings(talker, self.options[0][talker_name])
    except ct_api.TalkerError as e:
        # Remove talker as we failed to apply the settings
        del self.talker_plugins[e.source]
        logger.exception("Failed to initialize talker settings. Error %s", e)
```

Which can entirely be replaced with this ignoring the `set_talker_settings` function:
```python
for talker_name, talker in list(self.talker_plugins.items()):
    try:
        talker.parse_settings(self.options[0][talker_name])
    except Exception as e:
        # Remove talker as we failed to apply the settings
        del self.talker_plugins[talker_name]
        logger.exception("Failed to initialize talker settings. Error %s", e)
```

mizaki replied:
I think it's where it was leftover and I didn't think it through.

lordwelch commented on comictaggerlib/ctoptions/talker_plugins.py:
```diff
@@ -0,0 +1,18 @@
+def register_talker_settings(parser: settngs.Manager, plugins: Mapping[str, ComicTalker]) -> None:
+    for talker_name, talker in plugins.items():
+        try:
+            parser.add_group(talker_name, talker.register_settings, False)
```
I want all of these groups to have `talker_` as a prefix in the group name. I don't want it to be possible to have a name conflict with a ComicTagger group. If two talkers have a conflict they can deal with it themselves.

lordwelch commented on comictaggerlib/main.py:
```diff
@@ -136,32 +151,14 @@ def main(self) -> None:
-            if self.config_load_success:
-                self.manager.save_file(self.options[0], settings_path)
-
-        if self.options[0].commands_only_set_cv_key:
-            if self.config_load_success:
-                print("Key set")  # noqa: T201
-                return
```
This should not be removed

lordwelch commented on comictaggerlib/main.py:
```diff
@@ -136,32 +151,14 @@ def main(self) -> None:
-            error = (str(e), True)
+            talker_api = self.talker_plugins[self.options[0].talkers_source]
+        except Exception as e:
+            logger.exception(f"Unable to load talker {self.options[0].talkers_source}. Error: {str(e)}")
```
remove the `str(e)`

lordwelch commented on comictaggerlib/main.py:
```diff
@@ -136,32 +151,14 @@ def main(self) -> None:
+        except Exception as e:
+            logger.exception(f"Unable to load talker {self.options[0].talkers_source}. Error: {str(e)}")
+            # TODO error True can be changed to False after the talker settings menu generation is in
+            error = (f"Unable to load talker {self.options[0].talkers_source}. Error: {str(e)}", True)
```
remove the `str(e)`

lordwelch commented on comictalker/talkerbase.py:
```diff
@@ -137,27 +138,21 @@ def __init__(self, source: str = "", sub_code: int = 0, desc: str = "") -> None:
-        # joinurl only works properly if there is a trailing slash
-        if tmp_url.path and tmp_url.path[-1] != "/":
-            tmp_url = tmp_url._replace(path=tmp_url.path + "/")
+    def register_settings(self, parser: settngs.Manager) -> None:
+        """Allows registering settings using the settngs package with an argparse like interface"""
```
Add a `return None`

lordwelch commented on comictalker/talkerbase.py:
```diff
@@ -137,27 +138,21 @@ def __init__(self, source: str = "", sub_code: int = 0, desc: str = "") -> None:
+        """Allows registering settings using the settngs package with an argparse like interface"""
 
-        self.api_url = tmp_url.geturl()
+    def parse_settings(self, settings: dict[str, Any]) -> None:
+        """settings is a dictionary of options defined in register_settings.
+        It is only guaranteed that the settings defined in register_settings will be present."""
```
Add a `return None`

mizaki committed:
Merge branch 'develop' into talker_settings

# Conflicts:
#	comictaggerlib/ctoptions/__init__.py
#	comictaggerlib/main.py
#	comictalker/talkers/comicvine.py

mizaki committed:
After merge isort

mizaki committed:
Simplify parse_settings. Prefix talker_ to group name. Add back setting CV key via commandline. Other small changes as requested.

mizaki committed:
Add series match threshold back into search_for_series as it is no longer available via the talkers own settings.
